src/System.CommandLine.Tests/Invocation/CommandHandlerTests.cs

```diff
@@ -507,5 +507,101 @@ public class OverridenVirtualTestCommandHandler : VirtualTestCommandHandler
             public override Task<int> InvokeAsync(InvocationContext context)
                 => Task.FromResult(41);
         }
+
+        [Fact]
+        public static void FromBindingContext_forwards_invocation_to_bound_handler_type()
+        {
+            var command = new RootCommand
+            {
+                Handler = CommandHandler.FromBindingContext<BindingContextResolvedCommandHandler>()
+            };
+            command.Handler.Should().NotBeOfType<BindingContextResolvedCommandHandler>();
+            var parser = new CommandLineBuilder(command)
+                .ConfigureBindingContext(context => context.AddService<BindingContextResolvedCommandHandler>())
+                .Build();
+
+            var console = new TestConsole();
+            parser.Invoke(Array.Empty<string>(), console);
+            console.Out.ToString().Should().Be(typeof(BindingContextResolvedCommandHandler).FullName);
+        }
+
+        [Fact]
+        public static void Subsequent_call_to_configure_overrides_service_registration()
+        {
+            BindingContextCommandHandlerAction action = (handler, Console) =>
+            {
+                handler.Should().BeOfType<BindingContextCommandHandler2>();
+            };
+            var parser = new CommandLineBuilder(new RootCommand
+            {
+                Handler = CommandHandler.FromBindingContext<IBindingContextCommandHandlerInterface>()
+            })
+                .ConfigureBindingContext(context => context.AddService(_ => action))
+                .ConfigureBindingContext(context => context.AddService<IBindingContextCommandHandlerInterface, BindingContextCommandHandler1>())
+                .ConfigureBindingContext(context => context.AddService<IBindingContextCommandHandlerInterface, BindingContextCommandHandler2>())
```

Comment:
One concern I have with this approach is that all handler registrations for all subcommands will happen regardless of which subcommand was used, when only one subcommand is ever used. This can be a bit wasteful and is part of why we've generally steered away from the common pattern for longer-lived apps where DI is configured up front but the cost is amortized over a longer lifetime. This is very often not the case for command-line apps.

Reply:
@jonsequitur In this case (using the lazily bound registration) on the registration is added, no instance of `BindingContextCommandHandler1` nor `BindingContextCommandHandler2` is actually created at this point. And neither does the call to `CommandHandler.FromBindingContext`. The actual instance is first created when the command handler for the choses command is invoked.

Reply:
Understood. At invocation time though, all of these registrations happen on the `BindingContext`, and coupled with the addition of non-lazy `AddService` calls, this encourages the pattern I mentioned above.